File: C++/Period.cpp
// reference:
// H.M.Edwards "Fermat's Last Theorem" section 4.5

// uses NTL
//   http://www.shoup.net/ntl

#include "Period.h"
#include<exception>
using namespace NTL;

long Period::f;// length of period
long Period::e;// number of conjugates
Mat<long> Period::W;// multiplication table
ZZX Period::MPoly;// minimal polynomial

void Period::init(long f_)
// set length f of period.
// Assume CYC::init(n) has been executed.
// Assume f divides n-1.
{
    long i,j,k,n(CYC::n),n1(n-1);
    if(n1%f_)
        throw std::runtime_error("wrong f in Period");
    f = f_;
    e = n1/f;

    // minimal polynomial
    ZZX eta;
    for(i=0; i<n1; i+=e)
        SetCoeff(eta, CYC::g_pow[i]);
    eta %= CYC::modulus;
    MinPolyMod(MPoly, eta, CYC::modulus);

    // multiplication table
    // eta_0*eta_i = sum_{j=0}^{e-1} W[0][j]*eta_j
    W.SetDims(e,e);
    for(i=0; i<e; i++) {
        for(j=0; j<e; j++) W[i][j] = 0;
        for(j=i; j<n1; j+=e) {
            k = (1 + CYC::g_pow[j])%n;
            if(k) W[i][CYC::log_g[k]%e]++;
            else for(k=0; k<e; k++) W[i][k] -= f;
        }
    }
}

void conv(CYC& b, const Period& a)
// transform a to cyclotomic integer b
// as a polynomial in omega = exp(2pi*i/n)
{
    long i,j,n1(CYC::n-1);
    b.SetLength(CYC::n);
    clear(b[0]);
    for(i=0; i<a.e; i++)
        for(j=i; j<n1; j+=a.e)
            b[CYC::g_pow[j]] = a[i];
    b %= CYC::modulus;
}

void mul(Period& c, const Period& a, const Period& b)
// c = a*b
{
    if(&c==&a) { Period d(a); mul(c,d,b); return; }
    if(&c==&b) { Period d(b); mul(c,a,d); return; }
    long i,j,k;
    ZZ u;
    clear(c);
    for(i=0; i<a.e; i++) {
        for(j=0; j<b.e; j++) {
            mul(u, a[i], b[j]);
            for(k=0; k<c.e; k++)
                MulAddTo(c[k], u, c.MulTab(i,j,k));
        }
    }
}

Period& operator*=(Period& b, const Period& a)
{ mul(b,b,a); return b; } // b *= a

long IsRational(ZZ& x, const Period& a)
// test if a is rational
// return 1 and set x=a[0]-a[1] if a is rational
// return 0 if a is not rational (x is unchanged)
// 
{
    for(long i=0; i<a.e; i++)
        if(a[i] != a[0]) return 0;
    negate(x, a[0]); return 1;
}

File: Period.py
# reference:
# H.M.Edwards "Fermat's Last Theorem" section 4.5

from CYC import *
from sympy.abc import x

class Period:
    """ cyclotomic periods
    represented by array of int of length e
    """
    @staticmethod
    def init(f):
        """ set length f of period
        Assume CYC.init(n) has been executed.
        Assume f divides n-1
        """
        n1 = CYC.n1
        e,r = divmod(n1,f) # e = number of conjugates
        if r:
            raise RuntimeError("wrong f in Period")
        r = range(e)
        i = [CYC.g_pow[np.r_[i:n1:e]] for i in r]
        j = [np.roll(np.r_[:e], -i) for i in r]
        Period.e, Period.f = e,f
        Period.index = np.asarray(i)
        Period.cjind = np.asarray(j)

        # multiplication table
        w = np.zeros((e,e), dtype='object')
        for i in r:
            for j in range(f):
                k = (1 + Period.index[i,j])%CYC.n
                if k: w[i, CYC.log_g[k]%e] += 1
                else: w[i] -= f

        w = [np.roll(w,i,axis=(0,1)) for i in r]
        Period.w = np.asarray(w, dtype='object')
        # eta_i*eta_j = sum_{k=0}^{e-1} w[i,j,k]*eta_k

        # minimal polynomial
        u,a,y = [], 1, -Period.basis(0)
        for i in r:
            u.append(y)
            for j in range(i,-1,-1):
                if j<i: u[j] *= y
                if j: u[j] += u[j-1]
            y = y.conj()

        for c in u[::-1]: a = a*x + c.to_int()
        Period.MPoly = sp.expand(a)
        # (x-eta_0)(x-eta_1)...(x-eta_{e-1})

    @staticmethod
    def basis(i=0):
        """ return eta_i (i=0,...,e-1) """
        return Period(np.roll([1]+[0]*(Period.e-1), i))

    def __init__(a,c):
        """ set coefficients c of periods
        as sum_{j=0}^{e-1} c_j eta_j
        """
        a.c = np.asarray(c, dtype='object')
        a.c.resize(a.e)

    def __neg__(a): return Period(-a.c) # -a

    def __add__(a,b): # a+b, b must be Period object
        return Period(a.c + b.c)

    def __mul__(a,b): # a*b, b must be Period object
        c = np.dot(a.c, np.dot(b.c, b.w))
        return Period(c)

    # conjugate by cyclic permutation: eta_j -> eta_{j+i}
    def conj(a, i=1): return Period(a.c[a.cjind[i]])

    # product of all conjugates of a
    def norm(a):
        b = [a.conj(i) for i in range(a.e)]
        return np.prod(b, axis=0).to_int()

    # -(coeff of eta_0)
    def to_int(a): return -a.c[0]

    def isRational(a):
        """ test if a is rational.
        return None if a is not rational,
        return a.to_int() if a is rational.
        """
        if np.all(a.c == a.c[0]):
            return a.to_int()
        else: return None

    def to_CYC(a):
        """ transform a to cyclotomic integer
        as a polynomial in omega = exp(2pi*i/n)
        """
        y = np.zeros(CYC.n, dtype='object')
        y[a.index] = a.c[:,np.newaxis]
        return CYC(y - y[-1]) # normalize
